fints/types.py

```diff
@@ -0,0 +1,421 @@
+        stream.write( (prefix + level*indent) + "]){}\n".format(trailer) )
+
+    def find_segments(self, query=None, version=None, callback=None, recurse=True):
+        """Yields an iterable of all matching segments.
+
+        :param query: Either a str or class specifying a segment type (such as 'HNHBK', or :class:`~fints.segments.message.HNHBK3`), or a list or tuple of strings or classes.
+                     If a list/tuple is specified, segments returning any matching type will be returned.
+        :param version: Either an int specifying a segment version, or a list or tuple of ints.
+                        If a list/tuple is specified, segments returning any matching version will be returned.
+        :param callback: A callable that will be given the segment as its sole argument and must return a boolean indicating whether to return this segment.
+        :param recurse: If True (the default), recurse into SegmentSequenceField values, otherwise only look at segments in this SegmentSequence.
+
+        The match results of all given parameters will be AND-combined.
+        """
+
+        if query is None:
+            query = []
+        elif isinstance(query, str) or not isinstance(query, (list, tuple, Iterable)):
+            query = [query]
+
+        if version is None:
+            version = []
+        elif not isinstance(version, (list, tuple, Iterable)):
+            version = [version]
+
+        if callback is None:
+            callback = lambda s: True
+
+        for s in self.segments:
+            if ((not query) or any( (isinstance(s, t) if isinstance(t, type) else s.header.type == t) for t in query)) and \
+                ((not version) or any(s.header.version == v for v in version)) and \
+                callback(s):
+                yield s
+
+            if recurse:
+                for name, field in s._fields.items():
+                    val = getattr(s, name)
+                    if val and hasattr(val, 'find_segments'):
+                        yield from val.find_segments(query=query, version=version, callback=callback, recurse=recurse)
+
+    def find_segment_first(self, *args, **kwargs):
+        """Finds the first matching segment.
+
+        Same parameters as find_segments(), but only returns the first match, or None if no match is found."""
+
+        for m in self.find_segments(*args, **kwargs):
+            return m
+
+        return None
+
+    def find_segment_highest_version(self, query=None, version=None, callback=None, recurse=True, default=None, *args, **kwargs):
+        """Finds the highest matching segment.
+
+        Same parameters as find_segments(), but returns the match with the highest version, or default if no match is found."""
+        # FIXME Test
+
+        retval = None
+
+        for s in self.find_segments(query=query, version=version, callback=callback, recurse=recurse, *args, **kwargs):
```

Comment:
``.find_segments()`` doesn't take additional arguments, so this won't work. Would the correct thing be to remove the args/kwargs here or to add them in the other place?

Reply:
Remove. I think this happened when going from `def find_segment_highest_version(self, *args, **kwargs)` to a version where the new `default` parameter is present, but *after* the normal parameters that are passed through to `find_segments()`.